### DISCUSSION TITLE
The `reconcile` function unexpected behavior

### DISCUSSION THREAD

**Discussion by SonyStone** [solidjs/solid#1961] [Q&A]
### Describe the bug

I was working with the `createUndoHistory` function and the example suggested using the reconcile function to restore the previous state. But for some reason, in some cases the states were skipped. After a long digging in the code, I came to the conclusion that the reconcile function in some circumstances mutates the object passed to it, breaking the work of the `createUndoHistory` function.

I was able to reproduce the bug in stackblitz, hope it will help.

### Your Example Website or App

https://stackblitz.com/edit/typescript-fcz5m5?file=index.ts

### Steps to Reproduce the Bug or Issue

1. https://stackblitz.com/edit/typescript-fcz5m5?file=index.ts
2. Open the console
3. Seeing that the `first` object mutated and got the wrong value, when it shouldn't

### Expected behavior

Object `first` should not change

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by ryansolid** [comment]
I suspect I know what this is (although I haven't had a chance to look yet) and it comes down to the fact that since we are by reference we do need to mutate things we get a hold of. There is no other way to do it. There is an argument for deep cloning everything that gets mutated into the object but that is costly in its own right. If I was constantly going backwards I could see an argument for that but creation would be paying the price. The problem is less so with `reconcile` with how `createStore` works.

What I generally do for these sort of back/forward systems is clone on leaving the state. https://codesandbox.io/s/pkjw38r8mj?file=/src/index.js

I've also toyed with a `createImmutable` that doesn't have this behavior: https://codesandbox.io/s/createimmutable-8brhtk

**Comment by SonyStone** [comment]
![image](https://github.com/solidjs/solid/assets/8790196/d784d476-e412-48b1-840c-d82ce00cd520)
I'm not too sure about your answer, but in this example it's mutating an object that shouldn't be changing in any way at all.

And I'm not using any stores here yet

**Comment by ryansolid** [comment]
Well it is happening because when we use `start` it removes the value at property 0.. and then when you reconcile `first`, it sets the key at 0 to first entry. From that point on it is mutating the object that first has, so that the second to last reconcile actually changes first, and then the final one ends up with the same reference exits out.

Here are all the operations
```js
// reconcile(second)(structuredClone(third))
result[0].image = second[0].image

// reconcile(first)(result)
result[0].image = first[0].image

// reconcile(start)(result)
result[0] = undefined

// reconcile(first)(result)
result[0] = first[0] // can only match up to the 0 property

// reconcile(second)(result)
result[0].image = second[0].image // result[0] is first[0] so mutating first image property

// reconcile(first)(result)
result[0].image = first[0].image // no change its the same reference to the original second image property
```
The results here are what I expected as we rely on mutation. The first thought is to maybe deep clone everything coming in but then nothing will ever match. This is also expensive for people not doing reconcile with stores. Keep in mind this API exists purely for stores.

The alternative is to always clone and diff from the root. Which is why I made the createImmutable example above because the thing is Solid's stores can produce the same behavior even without `reconcile`. Like that same set of mutations above without `reconcile` would do the same thing. You would need the person doing the setting to do the cloning always from the root. Or I suppose the primitive could (like Immer style) but any change would require cloning the parents, which would require knowledge of parents which our current stores do not.

The reason is that fundamentally mutable stores can exist and update from multiple locations, as identity is not locked to a path. WIth immutable data if you change something at one path any other instances are still at the other value. In so we don't naturally have a single parent. I can think of ways to address that but they would be expensive.

I suppose this leaves the question of whether there should be two different store types, those that are immutable and those that are mutable, and from that perspective, the mutable has no business doing this sort of diffing and should only be doing key + merge based diffing. I think that is fair and probably should be considered. But it would be difficult for reconcile on a mutable store to act differently and be efficient for the primary case.

While this isn't a bug, I think it is worth further discussion. Instead of closing the issue I'm going to move it there.
